jujugui/static/gui/src/test/test_env_api.js

```diff
@@ -1298,6 +1299,222 @@ with this program.  If not, see <http://www.gnu.org/licenses/>.
+        // Perform the request.
+        const args = {
+          applicationName: 'wordpress',
+          charmURL: '',
+          channel: 'candidate',
+          resources: [{Name: 'myres', File: 'myres.tar.bz2'}]
+        };
+        env.addPendingResources(args, (err, ids) => {
+          assert.equal(err, 'invalid arguments: ' + JSON.stringify(args));
+          assert.deepEqual(ids, {});
+          done();
+        });
+      });
+
+      it('fails when no resources are provided', done => {
+        // Perform the request.
+        const args = {
+          applicationName: 'wordpress',
+          charmURL: 'wordpress-42',
+          resources: []
+        };
+        env.addPendingResources(args, (err, ids) => {
+          assert.equal(err, 'invalid arguments: ' + JSON.stringify(args));
+          assert.deepEqual(ids, {});
+          done();
+        });
+      });
+
+      it('fails when invalid resources are provided', done => {
+        // Perform the request.
+        env.addPendingResources({
+          applicationName: 'wordpress',
+          charmURL: 'wordpress-42',
+          resources: [{File: 'bad'}]
+        }, (err, ids) => {
+          assert.equal(err, 'resource without name: [{"File":"bad"}]');
+          assert.deepEqual(ids, {});
+          done();
+        });
+      });
+
+      it('handles global failures while adding resources', done => {
+        // Perform the request.
+        env.addPendingResources({
+          applicationName: 'redis',
+          charmURL: 'redis-0',
+          resources: [{Name: 'red', File: 'red.tgz'}]
+        }, (err, ids) => {
+          assert.strictEqual(err, 'bad wolf');
+          assert.deepEqual(ids, {});
+          done();
+        });
+        // Mimic response.
+        conn.msg({'request-id': 1, error: 'bad wolf'});
+      });
+
+      it('handles local failures while adding resources', done => {
+        // Perform the request.
+        env.addPendingResources({
+          applicationName: 'redis',
+          charmURL: 'redis-0',
+          resources: [{Name: 'red', File: 'red.tgz'}]
+        }, (err, ids) => {
+          assert.strictEqual(err, 'bad wolf');
+          assert.deepEqual(ids, {});
+          done();
+        });
+        // Mimic response.
+        conn.msg({
+          'request-id': 1,
+          response: {error: {message: 'bad wolf'}}
+        });
+      });
+
+      it('handles unexpected failures while adding resources', done => {
+        // Perform the request.
+        env.addPendingResources({
+          applicationName: 'redis',
+          charmURL: 'redis-0',
+          resources: [{Name: 'red', File: 'red.tgz'}]
+        }, (err, ids) => {
+          assert.strictEqual(err, 'unexpected response: {"pending-ids":[]}');
+          assert.deepEqual(ids, {});
+          done();
+        });
+        // Mimic response.
+        conn.msg({'request-id': 1, response: {'pending-ids': []}});
+      });
+    });
```

Comment:
In same cases the call fails and fires the callback even before actually calling the server. In those cases, as requests are not performed, simulating responses is not required.